### DISCUSSION TITLE
how-to-use-the-ioptions-pattern-for-configuration-in-asp-net-core-rc2/

### DISCUSSION THREAD

**Discussion by andrewlock** [andrewlock/blog-comments#141] [Announcements]
# how-to-use-the-ioptions-pattern-for-configuration-in-asp-net-core-rc2/

In this post I demonstrate the power of the IOptions<> pattern for configuration in ASP.NET Core, and describe a few of the problems you may run in to

https://andrewlock.net/how-to-use-the-ioptions-pattern-for-configuration-in-asp-net-core-rc2/

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at May 21 2016, 04:53, in reply to [Mark Vincze](https://disqus.com/by/markvincze/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/141#discussioncomment-5794927)</em>

---

<p>Thanks:) just FYI, I've updated the section on exposing interfaces as it's a little more nuanced than I realised at first!</p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at May 22 2016, 10:31, in reply to [Larry Ruckman](https://disqus.com/by/larry_ruckman/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/141#discussioncomment-5794930)</em>

---

<p>Ah excellent, that's good to know, cheers!</p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at August 18 2016, 04:16, in reply to [Vladimir Cheshko](https://disqus.com/by/vladimircheshko/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/141#discussioncomment-5794936)</em>

---

<p>Is I mentioned above, currently only string keys are supported. You could raise an issue on GitHub requesting support for this, or even better send a pull request! <a href="https://github.com/aspnet/Configuration" rel="nofollow noopener" title="https://github.com/aspnet/Configuration">https://github.com/aspnet/Configuration</a></p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at October 11 2016, 09:45, in reply to [jgaylord](https://disqus.com/by/jgaylord/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/141#discussioncomment-5794940)</em>

---

<p>There's not any major changes I'm aware of I'm afraid, so not sure where to tell you!</p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at October 19 2016, 08:19, in reply to [Tom Aalbers](https://disqus.com/by/tomaalbers/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/141#discussioncomment-5794953)</em>

---

<p>Glad it's been helpful:) I've answered the question on StackOverflow, hope it makes sense:)</p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at December 30 2016, 05:51, in reply to [cssface](https://disqus.com/by/cssface/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/141#discussioncomment-5794956)</em>

---

<p>You should be able to create an instance of IOptions&lt;mysettings&gt;() using:</p><p></p><pre><code><br>namespace Microsoft.Extensions.Options<br>IOptions&lt;mysettings&gt; value = Options.Create(new mysettings());<br></code></pre>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at February 10 2017, 11:24, in reply to [Saša Ćetković](https://disqus.com/by/scetkovic/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/141#discussioncomment-5794959)</em>

---

<p>Oops, yep discovered that's the way to do it and completely forgot I'd said that, thanks!</p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at March 20 2017, 05:29, in reply to [Nicholas Fulton](https://disqus.com/by/nicholasfulton/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/141#discussioncomment-5794961)</em>

---

<p>Yes, you can inject IConfiigurationRoot into your classes, and access it directly, but you need to register it with the DI service first.</p><p>Check out this post from Rick Strahl on how to do it: <a href="https://weblog.west-wind.com/posts/2016/may/23/strongly-typed-configuration-settings-in-aspnet-core#RefreshingConfigurationValues" rel="nofollow noopener" title="https://weblog.west-wind.com/posts/2016/may/23/strongly-typed-configuration-settings-in-aspnet-core#RefreshingConfigurationValues">https://weblog.west-wind.com/posts/2016/may/23/strongly-typed-configuration-settings-in-aspnet-core#RefreshingConfigurationValues</a></p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at October 09 2017, 07:07, in reply to [marco](https://disqus.com/by/disqus_3SbF0z6X35/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/141#discussioncomment-5794975)</em>

---

<p>You'll need to provide more details I'm afraid. I suggest you post a minimal example on Stack Overflow</p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at September 27 2018, 07:56, in reply to [am](https://disqus.com/by/disqus_kotMQ7NydX/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/141#discussioncomment-5794984)</em>

---

<p>This post is pretty old now, but it should still work. Have you looked at the section on models that won't bind: <a href="https://andrewlock.net/how-to-use-the-ioptions-pattern-for-configuration-in-asp-net-core-rc2/#models-that-won-t-bind" rel="nofollow noopener" title="https://andrewlock.net/how-to-use-the-ioptions-pattern-for-configuration-in-asp-net-core-rc2/#models-that-won-t-bind">https://andrewlock.net/how-to-use-the-ioptions-pattern-for-configuration-in-asp-net-core-rc2/#models-that-won-t-bind</a> ? If that doesn't help, you'll prob have good luck posting on StackOverflow with a sample of your source code and configuration.</p>

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at September 27 2018, 08:30, in reply to [am](https://disqus.com/by/disqus_kotMQ7NydX/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/141#discussioncomment-5794987)</em>

---

<p>It suggests your configuration isn't being loaded correctly rather than the binding. As I say, it's best to post on StackOverflow with an example that reproduces the problem - you'll get a quick answer I'm sure :)</p>

[a comment by an automated account is not shown]

**Reply by andrewlock** [reply]
<em>[Sock](https://disqus.com/by/andrewlockdotnet/) (@andrewlock) commented [on Disqus](https://disqus.com/home/forum/andrewlock/) at December 10 2018, 08:21, in reply to [precogtyrant](https://disqus.com/by/precogtyrant/)'s [comment](https://github.com/andrewlock/blog-comments/discussions/141#discussioncomment-5794993)</em>

---

<p>Thanks!</p>

[a comment by an automated account is not shown]

**Comment by emirhankurtulus** [comment]
Thank you. You helped me a lot.

**Comment by PhilipeRLeal** [comment]
Dear Andrew Lock.

I would like to thank you for your very explanatory and detailed post. 

One thing that got me wondering was how to access the settings that are stored in the appsettings.json file of the main project (e.g., the application layer) from another Project of the same solution?

If, for instance, we try to inject that IOptions<MyBindableAppDataClass> into another class that is implemented in a project that is not the application layer, I bet that this IOptions would try to retrieve information from an appsettings.json that may (or may not be) be available at this project in which this class is implemented. Am I right to assume that? If so, how can one counter-maneuver this problem?
